a.go

```diff
@@ -95,9 +125,14 @@ func (a *A) Failed() bool {
 
 // Fail marks the function as having failed but continues execution.
 func (a *A) Fail() {
+	var call func()
 	a.mu.Lock()
 	a.failed = true
+	call = a.failedCall
```

Comment:
I do not think this need to be inside the lock. `a.failedCall` does not look to be ever mutated.

Probably this will work.


```go
func (a *A) Fail() {
	if a.failedCall != nil {
		a.failedCall()
		return
	}
	a.mu.Lock()
	a.failed = true
	a.mu.Unlock()
}
```

Reply:
I think we just need to change the call in places so that the flag is set on both the copy and the parent.